src/Confluent.Kafka/Producer.cs

```diff
@@ -671,7 +673,41 @@ internal Producer(ProducerBuilder<TKey, TValue> builder)
             }
 
             this.ownedKafkaHandle = SafeKafkaHandle.Create(RdKafkaType.Producer, configPtr, this);
-            configHandle.SetHandleAsInvalid(); // config object is no longer usable.
+            configHandle.SetHandleAsInvalid();  // ownership was transferred.
+                                                // configPtr remains valid over the lifetime of this method.
+
+            Func<PartitionerDelegate, IntPtr> createTopicConfigWithPartitioner = (partitioner) => {
+                var topicConfigPtr = SafeTopicConfigHandle.DuplicateDefaultTopicConfig(configPtr);
+                Librdkafka.topic_conf_set_partitioner_cb(configPtr,
+                    (IntPtr rkt, IntPtr keydata, UIntPtr keylen, int partition_cnt, IntPtr rkt_opaque, IntPtr msg_opaque) =>
+                    {
+                        unsafe
+                        {
+                            var topicNamePtr = Librdkafka.topic_name(rkt);
```

Comment:
What's the performance implication of doing a alloc-copy-utf8transform for each partitioned message?
E.g., what's the thruput drop for 100 byte topics going at full speed?

Reply:
I don't know. we don't have a choice in the default all-topic partitioner case. if it's slow someone will complain and we can put energy into it then maybe. I don't reckon it is slow. the alloc will be cheap because it will be short-lived.

Reply:
Maybe you can compare throughput with and without this alloc, use the mock broker.

Reply:
prioritization! i would rather work on other things. this is getting stuck in the details when we have bigger stuff to do. waste of time

Reply:
Knowing the performance of a feature is part of the deliverable. Even though this particular case here with a string object instantiation might not be that interesting, you/we should know the throughput numbers with and without custom partitioners.